operator/src/main/resources/scripts/stop-server.py

```diff
@@ -2,23 +2,200 @@
 # Licensed under the Universal Permissive License v 1.0 as shown at
 # http://oss.oracle.com/licenses/upl.
 
+# This the the preStop hook script (stopServer.sh) that is specified by the domain resource.
+# This script will be called by Kubernetes before it kills a pod.  Kuberentes will kill
+# the pod once this script returns OR when the grace period timeout expires
+# (see Operator shutdown.timeoutSeconds)
 #
-# This is a WLST script for starting WL server via the node manager.
+# This script shuts down the local server running in the container (admin or managed).
+# There are 2 main scenarios, a domain with a Coherence cluster and one without.
+# If there is no Coherence cluster, then the code will attempt to connect to the
+# local server and do a shutdown.  If that doesn't work it will attempt to shutdown
+# using NodeManager.  If that fails it just exits and kubernetes will kill the pod.
+# If a Coherence cluster exists, then we must wait for the Coherence services to be
+# safe before the server is shutdown, or we risk losing data.  This requires a
+# a connection to the admin server and inspection of the Coherence MBeans.  The code
+# will not shutdown the server until Coherence is safe.  If Coherence never becomes safe,
+# then, eventually, the grace period will expire and Kubernetes will kill the pod.
 #
 # It users key and data files that were generated by the introspector
 # for its nmConnect credentials.
 #
 
 import sys
+import traceback
 import base64
+import time as systime
 
+# Get an ENV var
 def getEnvVar(var):
   val=os.environ.get(var)
   if val==None:
     print "ERROR: Env var ",var, " not set."
     sys.exit(1)
   return val
 
+
+# Connect to Node Manager and shut down the local server
+#
+def shutdownUsingNodeManager(domainName, domainDir):
+  try:
+    print('Shutdown: Attempting shutdown using NodeManager')
+    nmConnect(userConfigFile='/weblogic-operator/introspector/userConfigNodeManager.secure',
+              userKeyFile='/tmp/userKeyNodeManager.secure.bin',
+              host='127.0.0.1',port='5556',
+              domainName=domainName,
+              domainDir=domainDir,
+              nmType='plain')
+    print('Shutdown: Successfully connected to NodeManager')
+    nmKill(server_name)
+    print('Shutdown: Successfully killed server using NodeManager')
+  except Exception, e:
+    traceback.print_exc(file=sys.stdout)
+    print('Shutdown: Failed to kill server using NodeManager')
+    raise
+
+
+# Check if Coherence exists using wlst offline
+def doesCoherenceExist():
+  try:
+    readDomain(domain_path)
```

Comment:
I wonder how slow a 'readDomain' call can be.  I suspect it might be significantly slow.  We're trying to speedup server-lifecycle and every second counts...  Is  there another way to deduce Coherence is active?

Reply:
Yes I can go look at config.xml directly but I was trying to avoid that.  I can make that change if you think it is important.

Reply:
Marking resolved since the change was made.